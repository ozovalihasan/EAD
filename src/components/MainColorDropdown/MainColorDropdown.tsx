import useCustomizationStore from '@/zustandStore/customizationStore';
import { memo, useRef } from 'react';
import { availableColors, availableColorsType} from '@/helpers';
import { handleMouseLeaveForSelect, handleMouseUpForSelect } from '@/helpers';

export const MainColorDropdown = memo(() => {
  
  const selectedColor = useCustomizationStore(store => store.mainColor)
  const changeMainColor = useCustomizationStore(store => store.changeMainColor)

  const selectEl = useRef<HTMLSelectElement | null>(null);

  
  return (
    <div 
      className='w-full relative  whitespace-nowrap'
      onMouseLeave={() => handleMouseLeaveForSelect(selectEl)}
    >
      <div 
        className='truncate p-2 rounded-md capitalize custom-select-button'
        onMouseUp={() => handleMouseUpForSelect(selectEl)} 
      >
        Main Color: { selectedColor }
      </div>
      <select
        ref={selectEl}
        className="hidden absolute left-0 bottom-full z-10 w-full custom-select-options"
        value={selectedColor}
        onChange={(selectedOption) => {
          changeMainColor(selectedOption.target.value as availableColorsType)}
        } 
        title="Select a main color"
        size={availableColors.length }
      >
        {availableColors.map((color: string) => (
          <option
            key={color}
            value={color}
            className="p-2 truncate capitalize"
            title={color}
          >
            {color}
          </option>
        ))}

      </select>
    </div>

  )
})