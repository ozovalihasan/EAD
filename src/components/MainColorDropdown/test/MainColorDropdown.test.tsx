import { MainColorDropdown } from '../MainColorDropdown';
import { render, screen, renderHook, fireEvent } from "@testing-library/react";
import useCustomizationStore from '@/zustandStore/customizationStore';

let renderReadyComponent: JSX.Element;

beforeEach(() => {
    
useCustomizationStore.setState({ 
    locationSidebar: "right",
    showTextOnEdges: true,
    mainColor: "sky",
    changeMainColor: jest.fn(),
    toggleLocationSidebar: jest.fn(),
    toggleTextMode: jest.fn(),
  })
    
  renderReadyComponent = (
    <MainColorDropdown />
  );
});

describe('<MainColorDropdown />', () => {

  it('calls changeTableSuperClass function of the store when the selected option is changed', () => {

    render(renderReadyComponent );
    const { result } = renderHook(() => useCustomizationStore());

    const selectEl = screen.getByTitle(/Select a main color/)
    fireEvent.change( selectEl, {target: {value: 'sky'}})
    expect(result.current.changeMainColor).toHaveBeenCalledTimes(1);
  });
  
  it('toggles the "hidden" class when mouseUp is fired', () => {

    render( renderReadyComponent );

    const selectEl = screen.getByTitle(/Select a main color/);
    const buttonEl = screen.getAllByText(/^Main Color: sky$/i)[0];

    fireEvent.mouseUp( buttonEl );
    
    expect(selectEl.classList).not.toContain("hidden");

    fireEvent.mouseUp( buttonEl );
    
    expect(selectEl.classList).toContain("hidden");

  });
  
  it('adds the "hidden" class when the mouse leaves a defined element ', () => {

    render( renderReadyComponent );

    const selectEl = screen.getByTitle(/Select a main color/);
    selectEl.classList.remove("hidden")

    expect(selectEl.classList).not.toContain("hidden");

    fireEvent.mouseLeave( selectEl );

    expect(selectEl.classList).toContain("hidden");

  });

  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });

});