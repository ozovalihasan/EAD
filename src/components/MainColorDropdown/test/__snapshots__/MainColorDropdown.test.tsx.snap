// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainColorDropdown /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-64 relative  whitespace-nowrap"
      >
        <div
          class="truncate p-2 rounded-md capitalize custom-select-button"
        >
          Main Color: 
          sky
        </div>
        <select
          class="hidden absolute left-0 top-full w-full custom-select-options"
          size="1"
          title="Select a main color"
        >
          <option
            class="p-2 truncate capitalize"
            title="sky"
            value="sky"
          >
            sky
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-64 relative  whitespace-nowrap"
    >
      <div
        class="truncate p-2 rounded-md capitalize custom-select-button"
      >
        Main Color: 
        sky
      </div>
      <select
        class="hidden absolute left-0 top-full w-full custom-select-options"
        size="1"
        title="Select a main color"
      >
        <option
          class="p-2 truncate capitalize"
          title="sky"
          value="sky"
        >
          sky
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
